ddsc/core/remotestore.py

```diff
@@ -0,0 +1,323 @@
+
+    def _add_child(self, parent, child):
+        """
+        Add file or folder(child) to parent.
+        :param parent: RemoteProject/RemoteFolder to add child to
+        :param child: dict JSON data back from remote store
+        """
+        kind = child['kind']
+        if kind == KindType.folder_str:
+            parent.add_child(self._read_folder(child))
+        elif kind == KindType.file_str:
+            parent.add_child(self._read_file_metadata(child))
+        else:
+            raise ValueError("Unknown child type {}".format(kind))
+
+    def _read_folder(self, folder_json):
+        """
+        Create RemoteFolder and query it's children.
+        :param folder_json: dict JSON data back from remote store
+        :return: RemoteFolder folder we filled in
+        """
+        folder = RemoteFolder(folder_json)
+        response = self.data_service.get_folder_children(folder.id, '').json()
+        for child in response['results']:
+            self._add_child(folder, child)
+        return folder
+
+    def _read_file_metadata(self, file_json):
+        """
+        Create RemoteFile based on file_json and fetching it's hash.
+        :param file_json: dict JSON data back from remote store
+        :return: RemoteFile file we created from file_json
+        """
+        remote_file = RemoteFile(file_json)
+        response = self.data_service.get_file(remote_file.id)
+        file_hash = response.json()['upload']['hash']
+        if file_hash:
+            remote_file.set_hash(file_hash['value'], file_hash['algorithm'])
+        return remote_file
+
+    def lookup_user_by_email_or_username(self, email, username):
+        if username:
+            return self.lookup_user_by_username(username)
+        else:
+            return self.lookup_user_by_email(email)
+
+    def lookup_user_by_name(self, full_name):
+        """
+        Query remote store for a single user with the name full_name or raise error.
+        :param full_name: str Users full name separated by a space.
+        :return: RemoteUser user info for single user with full_name
+        """
+        res = self.data_service.get_users_by_full_name(full_name)
+        json_data = res.json()
+        results = json_data['results']
+        found_cnt = len(results)
+        if found_cnt == 0:
+            raise ValueError("User not found:" + full_name)
+        elif found_cnt > 1:
+            raise ValueError("Multiple users with name:" + full_name)
+        user = RemoteUser(results[0])
+        if user.full_name.lower() != full_name.lower():
+            raise ValueError("User not found:" + full_name)
+        return user
+
+    def lookup_user_by_username(self, username):
+        """
+        Finds the single user who has this username or raises ValueError.
+        :param username: str username we are looking for
+        :return: RemoteUser user we found
+        """
+        matches = [user for user in self.fetch_all_users() if user.username == username]
```

Comment:
Matter of style, but `filter` might be more expressive than a comprehension here